src/main/java/com/google/devtools/build/lib/exec/SpawnLogReconstructor.java

```diff
@@ -13,38 +13,106 @@
 
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
 import com.github.luben.zstd.ZstdInputStream;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
 import com.google.common.collect.ImmutableList;
 import com.google.devtools.build.lib.exec.Protos.ExecLogEntry;
 import com.google.devtools.build.lib.exec.Protos.File;
 import com.google.devtools.build.lib.exec.Protos.SpawnExec;
-import com.google.devtools.build.lib.util.Pair;
 import com.google.devtools.build.lib.util.io.MessageInputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.util.ArrayDeque;
+import java.util.ArrayList;
 import java.util.Collection;
 import java.util.HashMap;
 import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
 import java.util.Map;
 import java.util.SortedMap;
 import java.util.SortedSet;
 import java.util.TreeMap;
 import java.util.TreeSet;
+import java.util.function.Consumer;
+import java.util.regex.MatchResult;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Stream;
 import javax.annotation.Nullable;
 
 /** Reconstructs an execution log in expanded format from the compact format representation. */
 public final class SpawnLogReconstructor implements MessageInputStream<SpawnExec> {
+  // Examples:
+  // * bazel-out/k8-fastbuild/bin/pkg/file.txt (repo: null, path: "pkg/file.txt")
+  // * bazel-out/k8-fastbuild/bin/external/some_repo/pkg/file.txt (repo: "some_repo", path:
+  // "pkg/file.txt")
+  private static final Pattern DEFAULT_GENERATED_FILE_RUNFILES_PATH_PATTERN =
+      Pattern.compile("(?:bazel|blaze)-out/[^/]+/[^/]+/(?:external/(?<repo>[^/]+)/)?(?<path>.+)");
+
+  // Examples:
+  // * pkg/file.txt (repo: null, path: "pkg/file.txt")
+  // * external/some_repo/pkg/file.txt (repo: "some_repo", path: "pkg/file.txt")
+  private static final Pattern DEFAULT_SOURCE_FILE_RUNFILES_PATH_PATTERN =
+      Pattern.compile("(?:external/(?<repo>[^/]+)/)?(?<path>.+)");
+
+  // Examples:
+  // * bazel-out/k8-fastbuild/bin/pkg/file.txt (repo: null, path: "pkg/file.txt")
+  // * bazel-out/some_repo/k8-fastbuild/bin/pkg/file.txt (repo: "some_repo", path: "pkg/file.txt")
+  // * bazel-out/k8-fastbuild/k8-fastbuild/bin/pkg/file.txt (repo: "k8-fastbuild", path:
+  // "pkg/file.txt")
+  //
+  // Repo names are distinguished from mnemonics via a positive lookahead on the following segment,
+  // which in the case of a repo name is a mnemonic and thus contains a hyphen, whereas a mnemonic
+  // is followed by an output directory name, which does not contain a hyphen unless it is
+  // "coverage-metadata" (which in turn is not likely to be a mnemonic).
+  private static final Pattern SIBLING_LAYOUT_GENERATED_FILE_RUNFILES_PATH_PATTERN =
+      Pattern.compile(
+          "(?:bazel|blaze)-out/(?:(?<repo>[^/]+(?=/[^/]+-[^/]+/)(?!/coverage-metadata/))/)?[^/]+/[^/]+/(?<path>.+)");
```

Comment:
This is very ugly, but I don't currently see a way around it. I hope that the namespace conflict between repo names and configs can be resolved before `--experimental_sibling_repository_layout` is flipped.

Note that in Bazel 9 and later, we could instead check for the presence of `+` in repo names and prohibit it in config mnemonics.

Reply:
Yeah, I don't see any good alternatives either, and, since it currently looks like `--experimental_sibling_repository` layout won't ever be flipped, this might be forever :( Oh well.